From aa0514464238524294cc9f9fb713af1c36208cf9 Mon Sep 17 00:00:00 2001
From: Yumi Yukimura <me.cafebabe@gmail.com>
Date: Thu, 24 Jul 2025 00:47:44 +0800
Subject: [PATCH 3/5] tinyhal: configmgr: Add option to continue on ctl open
 error

* Useful for generic targets where we want to define some common
  ctls that may not present on all devices

Change-Id: I7e2b481b247b9aab8dba2838bdeddcdbe06174f6
---
 configmgr/Android.bp     | 8 ++++++++
 configmgr/audio_config.c | 4 ++++
 2 files changed, 12 insertions(+)

diff --git a/configmgr/Android.bp b/configmgr/Android.bp
index c1da4a2..9f96df5 100644
--- a/configmgr/Android.bp
+++ b/configmgr/Android.bp
@@ -49,6 +49,7 @@ soong_config_module_type {
     module_type: "cc_defaults",
     config_namespace: "tinyhal",
     bool_variables: [
+        "libaudiohalcm_continue_on_ctl_open_error",
         "tinyalsa_no_add_new_ctrls",
         "tinyalsa_no_ctl_get_id",
     ],
@@ -57,6 +58,10 @@ soong_config_module_type {
     ],
 }
 
+soong_config_bool_variable {
+    name: "libaudiohalcm_continue_on_ctl_open_error",
+}
+
 soong_config_bool_variable {
     name: "tinyalsa_no_add_new_ctrls",
 }
@@ -68,6 +73,9 @@ soong_config_bool_variable {
 libaudiohalcm_cc_defaults {
     name: "libaudiohalcm-defaults",
     soong_config_variables: {
+        libaudiohalcm_continue_on_ctl_open_error: {
+            cppflags: ["-DLIBAUDIOHALCM_CONTINUE_ON_CTL_OPEN_ERROR"],
+        },
         tinyalsa_no_add_new_ctrls: {
             cppflags: ["-DTINYALSA_NO_ADD_NEW_CTRLS"],
         },
diff --git a/configmgr/audio_config.c b/configmgr/audio_config.c
index b936472..facf307 100644
--- a/configmgr/audio_config.c
+++ b/configmgr/audio_config.c
@@ -522,7 +522,11 @@ static void apply_ctls_l(struct config_mgr *cm, struct ctl *pctl, const int ctl_
 
     for (i = 0; i < ctl_count; ++i, ++pctl) {
         if (ctl_open(cm, pctl) != 0) {
+#ifdef LIBAUDIOHALCM_CONTINUE_ON_CTL_OPEN_ERROR
+            continue;
+#else
             break;
+#endif
         }
 
         ctl = ctl_get_ptr(cm, &pctl->ref);
-- 
2.39.5

